src/python/pants/source/source_root.py

```diff
@@ -72,40 +92,16 @@ def find_by_path(self, path: str) -> Optional[SourceRoot]:
          Silently creating source roots on the fly is confusing legacy behavior that we shouldn't
          carry over into v2.
         """
-        matched = self._trie.find(path)
-        if matched:
-            return matched
-        elif self._options.unmatched == "fail":
-            return None
-        # If no source root is found, use the path directly.
-        return SourceRoot(path)
+        try:
+            return self.strict_find_by_path(path)
+        except NoSourceRootError:
+            if self._fail_if_unmatched:
+                return None
+            # If no source root is found, use the path directly.
+            return SourceRoot(path)
 
     def traverse(self) -> Set[str]:
-        return self._trie.traverse()
-
-    def all_roots(self):
-        """Return all known source roots.
-
-        Returns a generator over SourceRoot instances.
-
-        Note: Requires a directory walk to match actual directories against patterns.
-        However we don't descend into source roots, once found, so this should be fast in practice.
-        Note: Does not follow symlinks.
-        """
-        project_tree = FileSystemProjectTree(get_buildroot(), self._options.pants_ignore)
-
-        fixed_roots = set()
-        for root in self._trie.fixed():
-            if project_tree.exists(root):
-                yield SourceRoot(root)
-            fixed_roots.add(root)
-
-        for relpath, dirnames, _ in project_tree.walk("", topdown=True):
-            match = self._trie.find(relpath)
-            if match:
-                if not any(fixed_root.startswith(relpath) for fixed_root in fixed_roots):
-                    yield match  # Found a source root not a prefix of any fixed roots.
-                del dirnames[:]  # Don't continue to walk into it.
+        return sorted(str(r) for r in self._roots) if self._roots else self._trie.traverse()  # type: ignore
```

Comment:
Please use the specific code for this ignore, like `# type: ignore[assignment]`. MyPy will show the code in the error message.

Reply:
Can do, but note that this is going away in 1.30.0.dev0.